File: Code/Java/DFS.java
import java.util.*;

public class DFS extends TraverseAlgorithm {

    public static void main(String[] args) throws Exception {
        DFS dfs = new DFS(args);
        dfs.run();
    }

    private Set<Integer> visited = new LinkedHashSet<Integer>();
    
    public DFS(String[] args) throws Exception {
        super(args);
    }    
    
    public DFS(Graph graph) {
        super(graph);
    }
    
    /*
    
      Pseudo code

        dfs(vertex v)
        {
        visit(v);
        for each neighbor w of v
            if w is unvisited
            {
            dfs(w);
            add edge vw to tree T
            }
        }

    */
    @Override
    public Boolean run(int from) throws Exception {
        visit(from);
        if (graph.ifTarget(from)) {
            return true;
        }
        for( int nbr : graph.getNeighbours(from) ) {
            if (!visited.contains(nbr)) {
                Boolean ifFound = run(nbr);
                if(ifFound)
                    return true;
            }
        }
        return false;
    }

    public void visit(int noden) {
        this.processNode(noden);
        visited.add(noden);
    }


}

File: Code/C/DFS.c
#include "TraverseAlgorithm.h"

typedef struct DFS DFS;
struct DFS {
    Graph *graph;
    bool *visited;
    FILE *_file;
    int visitedCount;
};

DFS* newDFS_malloc(Graph *graph) {
    DFS *dfs = malloc(sizeof(DFS));
    dfs->graph = graph;
    dfs->visited = calloc( graph->count, sizeof(bool) );
    dfs->visitedCount = 0;
    return dfs;
}

void destroyDFS_free(DFS *dfs) {
    free(dfs->visited);
    free(dfs);
}

void DFSVisit(DFS *dfs, int noden) {
    CityInformation *info = dfs->graph->nodes[noden - 1];
    char *template = "(City: %s, Pop: %i, Elevation: %.1f)\n";
    fprintf(dfs->_file, template, info->city, (int)info->population, info->elevation);
    printf(template, info->city, (int)info->population, info->elevation);
    dfs->visited[noden - 1] = true;
    dfs->visitedCount++;
}

/*

  Pseudo code

    dfs(from vertex v)
      visit(v)
      for each not visited neighbor w of v
        dfs(w)
        add edge vw to tree T

*/

bool _DFSRun(DFS *dfs, int from) {
    DFSVisit(dfs, from);
    if (GraphIfTarget(dfs->graph, from)) {
        return true;
    }
    ListNode *nbr = GraphGetNeighbours(dfs->graph, from);
    for( ; nbr != NULL; nbr = nbr->next) {
        if (dfs->visited[nbr->number - 1] == false) {
            bool ifFound = _DFSRun(dfs, nbr->number);
            if (ifFound)
                return true;
        }
    }
    return false;
}

void DFSRun(DFS *dfs) {
    _DFSRun(dfs, dfs->graph->source);
}

int RunDFSAlgorithm(Graph *graph, FILE *file) {
    DFS* dfs = newDFS_malloc(graph);
    dfs->_file = file;
    DFSRun(dfs);
    int result = dfs->visitedCount;
    destroyDFS_free(dfs);
    return result;
}

int
main(int argc, char **argv) {
    runTraverseAlgorithm(argc, argv, "DFS", &RunDFSAlgorithm);
}

